src/BSON/Value.c

```diff
@@ -0,0 +1,665 @@
+		/* If the detected type was int32 but the original type was int64, we
+		 * replace the detected type without throwing. */
+		if (intern->value.value_type == BSON_TYPE_INT32 && Z_LVAL_P(type) == BSON_TYPE_INT64) {
+			intern->value.value_type = BSON_TYPE_INT64;
+		} else {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Detected type %s does not match indicated type %s", php_phongo_bson_type_to_string(intern->value.value_type), php_phongo_bson_type_to_string(Z_LVAL_P(type)));
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static HashTable* php_phongo_value_get_properties_hash(phongo_compat_object_handler_type* object, zval* type, bool is_temp)
+{
+	php_phongo_value_t* intern;
+	HashTable*          props;
+
+	intern = Z_OBJ_VALUE(PHONGO_COMPAT_GET_OBJ(object));
+	PHONGO_GET_PROPERTY_HASH_INIT_PROPS(is_temp, intern, props, 2);
+
+	/* get_debug_info will pass its own type. If it isn't given, use the type constant */
+	if (!type) {
+		zval value_type;
+
+		ZVAL_LONG(&value_type, intern->value.value_type);
+		zend_hash_str_update(props, "type", sizeof("type") - 1, &value_type);
+	} else {
+		zend_hash_str_update(props, "type", sizeof("type") - 1, type);
+	}
+
+	{
+		zval value;
+
+		phongo_bson_value_to_zval(&intern->value, &value);
+		zend_hash_str_update(props, "value", sizeof("value") - 1, &value);
+	}
+
+	return props;
+}
+
+PHONGO_DISABLED_CONSTRUCTOR(MongoDB_BSON_Value)
+
+static PHP_METHOD(MongoDB_BSON_Value, fromPHP)
+{
+	php_phongo_value_t* intern;
+	zval                zv;
+	zval*               data;
+
+	PHONGO_PARSE_PARAMETERS_START(1, 1)
+	Z_PARAM_ZVAL(data);
+	PHONGO_PARSE_PARAMETERS_END();
+
+	object_init_ex(&zv, php_phongo_value_ce);
+	intern = Z_VALUE_OBJ_P(&zv);
+
+	if (!phongo_zval_to_bson_value(data, &intern->value)) {
+		/* Exception already thrown */
+		zval_ptr_dtor(&zv);
+		return;
+	}
+
+	RETURN_ZVAL(&zv, 1, 1);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getType)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	RETURN_LONG(intern->value.value_type);
+}
+
+static PHP_METHOD(MongoDB_BSON_Value, getValue)
+{
+	php_phongo_value_t* intern;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	phongo_bson_value_to_zval(&intern->value, return_value);
+}
+
+#define PHONGO_VALUE_CHECK_TYPE(bson_value, expected_type)     \
+	if (bson_value.value_type != expected_type) {              \
+		UNEXPECTED_BSON_TYPE_EXCEPTION(bson_value.value_type); \
+		return;                                                \
+	}
+
+static PHP_METHOD(MongoDB_BSON_Value, getArray)
+{
+	php_phongo_value_t* intern;
+	bson_t*             bson;
+
+	PHONGO_PARSE_PARAMETERS_NONE();
+
+	intern = Z_VALUE_OBJ_P(getThis());
+
+	PHONGO_VALUE_CHECK_TYPE(intern->value, BSON_TYPE_ARRAY);
+
+	bson = bson_new_from_data(intern->value.value.v_doc.data, intern->value.value.v_doc.data_len);
+
+	phongo_packedarray_new(return_value, bson, false);
```

Comment:
`phongo_bson_value_to_zval` uses `bson_init_static` to initialise a BSON object, then passes it to `phongo_packedarray_new` with `copy=true`. This could probably be refactored to use `bson_new_from_data` and use `copy=false`, in which case we should document that ownership of the BSON structure will move to the newly created instance.
I can create a ticket for refactoring this.

Reply:
Sorry for not following up here. A separate ticket SGTM to we don't lose track of this.